tests/raw_query/tests.py

```diff
@@ -180,6 +180,12 @@ def test_params(self):
         self.assertEqual(len(results), 1)
         self.assertIsInstance(repr(qset), str)
 
+    def test_like_param(self):
+        query = "SELECT * FROM raw_query_author WHERE first_name like 'J%'"
+        qset = Author.objects.raw(query)
```

Comment:
Do you have a better example of why this would be useful? This can already be achieved using

```suggestion
        query = "SELECT * FROM raw_query_author WHERE first_name like %s"
        qset = Author.objects.raw(query, ('J%',))
```

Reply:
Yes. I understand that it can be avoided in the way how end user use raw function.

But it is not always possible to change solution in a way so you can generate params for your sql.

Reply:
So the issue remains the same as described. I can't use None params for raw function.

Reply:
I think the usage of `WHERE first_name like 'J%'` in your ticket description makes it a bit confusing.

Are you also getting  a crash when doing `Author.objects.raw('SELECT * FROM raw_query_author')`?

Reply:
> Do you have a better example of why this would be useful?

The exmple is very simple and generic. Any trusted service sends raw-sql to to another trasted service (Django) to performe it and receive the results in return. (Unfortunatelly I can't be more specific)

The thing is everything works just perfect untill one day we have received a query with like.

> This can already be achieved in a safer way doing
for my case it is not that easy as you show. And I was surprissed that allowing None is not that comple fix  (all tests passed, extra-test added), so why not allow using None, as well as it is allowed to use None for psycopg2?

Reply:
> I think the usage of WHERE first_name like 'J%' in your ticket description makes it a bit confusing.

Oh... I see. I can update ticket discrition, with service-service description I've showed above.

> Are you also getting a crash when doing Author.objects.raw('SELECT * FROM raw_query_author')?

No.

> Is this due to the presence of % in the parameter which psycopg2 expects to be placeholder?

I think yes.

Reply:
From the [_Executing custom SQL directly_ section of the _Performing raw SQL queries_ documentation](https://docs.djangoproject.com/en/3.1/topics/db/sql/#executing-custom-sql-directly)

> Note that if you want to include literal percent signs in the query, you have to double them **in the case you are passing parameters**

So it seems that it was meant to work the way you described but regressed at some point.

I'm afraid making this change now will break backward compatibility though as some folks might have worked around this issue by doubling percentage signs at this point.

We should determine in which version of Django this regressed or if this ever worked as documented to take an informed decision here.

Reply:
@charettes quick question. Can't find a way how to fix `isort` fail, not even sure, what is it.

Reply:
I wouldn't worry to much about `isort` for now as there's more important questions to resolve before this can be merged (PR has to be opened against the `master` branch, ticket has to be accepted wrt to the backward compatibility concerns) but here's a link from the documentation https://docs.djangoproject.com/en/3.1/internals/contributing/writing-code/coding-style/#imports

Reply:
> I'm afraid making this change now will break backward compatibility though as some folks might have worked around this issue by doubling percentage signs even when not passing parameters.

I update PR for backward compatibility.

Let me know what you think?

`def raw(self, raw_query, params=(), translations=None, using=None):`

*as well as for other functions where params is using*

Reply:
> We should determine for how long the implementation has diverged from the documentation to take an informed decision here as some users might have simply worked around this limitation by always doubling percentage signs even when no parameters are provided over the years. If we were to simply switch back to the documented way of doing things we could silently break queries of the form

The updated patch with backwards compatibility should prevent double-percentage being broken, and still allow using None as params